### PR TITLE
feat(misconf): loading embedded checks as a fallback

### PR DESCRIPTION
**Pull request by nikpivkin**
## Description

This PR adds functionality to fallback to embedded checks if an error occurs compiling an built-in check from a bundle. 

`main.tf`
```tf
resource "aws_s3_bucket" "test" {
  bucket = "test"
}

resource "aws_s3_bucket" "log_bucket" {
  bucket = "log_bucket"
}

resource "aws_s3_bucket_logging" "test" {
  bucket = aws_s3_bucket.test.id

  target_bucket = aws_s3_bucket.log_bucket.id
  target_prefix   = "log/"
}
```

1. `rm -rf ~/Library/Caches/trivy/policy`
2. `go run ./cmd/trivy conf main.tf -d --policy-bundle-repository ghcr.io/nikpivkin/trivy-policies:test`
```bash
2024-04-16T17:50:24+07:00       DEBUG   [misconf] 50:24.188902000 terraform.scanner.rego           Error occurred while parsing: Users/nikita/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3/enable_logging.rego, Users/nikita/Library/Caches/trivy/policy/content/policies/cloud/policies/aws/s3/enable_logging.rego:33: rego_type_error: undefined ref: bucket.logging.is_enabled.value
        bucket.logging.is_enabled.value
                       ^
                       have: "is_enabled"
                       want (one of): ["enabled" "targetbucket"]. 
Try loading embedded policy.
2024-04-16T17:50:24+07:00       DEBUG   [misconf] 50:24.189063000 terraform.scanner.rego           Found embedded policy: checks/cloud/aws/s3/enable_logging.rego
```

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/6505

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/iac/rego/load.go [hunk 1]
```diff
@@ -127,6 +145,60 @@ func (s *Scanner) LoadPolicies(enableEmbeddedLibraries, enableEmbeddedPolicies b
...
+		badPolicy, exists := s.policies[loc]
+		if !exists || badPolicy == nil {
+			continue
+		}
+
+		if !IsBuiltinNamespace(getModuleNamespace(badPolicy)) {
+			continue
+		}
+
+		s.debug.Log("Error occurred while parsing: %s, %s. \nTry loading embedded policy.", loc, e.Error())
```
Comment (simar7) [inline pkg/iac/rego/load.go:172]:
```suggestion
		s.debug.Log("Error occurred while parsing: %s, %s. Trying to fallback on embedded policy.", loc, e.Error())
```

File: pkg/iac/rego/load.go [hunk 2]
```diff
@@ -127,6 +145,60 @@ func (s *Scanner) LoadPolicies(enableEmbeddedLibraries, enableEmbeddedPolicies b
...
+
+		embedded := s.findMatchedEmbeddedCheck(badPolicy)
+		if embedded == nil {
+			s.debug.Log("Failed to find embedded policy: %s", loc)
+			continue
+		}
+
+		s.debug.Log("Found embedded policy: %s", embedded.Package.Location.File)
```
Comment (simar7) [inline pkg/iac/rego/load.go:180]: We should use the term `check` going forwards.
```suggestion
		s.debug.Log("Found embedded check: %s", embedded.Package.Location.File)
```

Comment (nikpivkin) [reply pkg/iac/rego/load.go]: Done https://github.com/aquasecurity/trivy/pull/6502/commits/e247ad23a09fa850d6387a280b0fe1a2e153a83f

File: pkg/iac/rego/load.go (hunk 1 above)
Comment (nikpivkin) [reply pkg/iac/rego/load.go]: Done https://github.com/aquasecurity/trivy/pull/6502/commits/e247ad23a09fa850d6387a280b0fe1a2e153a83f

**Review by knqyf263 — APPROVED** [review]
I just checked `commands/artifact` and `scanner/local`. They look good to me.

**Comment by simar7** [comment]
@nikpivkin one case I found was if we ever have a bad check, the fallback to embedded scenario will continue to take place until a new bundle is downloaded (which includes a fixed check). This would occur on each Trivy run.

Perhaps this is acceptable. I thought about re-downloading the bundle in this case (in addition to falling back for that particular run) but I think it's complicated and error prone in other ways.

**Review by simar7 — APPROVED** [review]

**Comment by nikpivkin** [comment]
@simar7 You mean re-download the previous version's bundle? Then all checks will be rolled back, including those with no errors.

File: pkg/iac/rego/load.go [hunk 3]
```diff
@@ -127,9 +144,73 @@ func (s *Scanner) LoadPolicies(enableEmbeddedLibraries, enableEmbeddedPolicies b
 	return s.compilePolicies(srcFS, paths)
 }
 
+func (s *Scanner) fallbackChecks(compiler *ast.Compiler) {
+
+	var excludedFiles []string
+
+	for _, e := range compiler.Errors {
+		loc := e.Location.File
```
Comment (chen-keinan) [inline pkg/iac/rego/load.go:152]: @nikpivkin @simar7 I have bump trivy-operator with latest trivy `0.51.1` and I'm getting `nil pointer` exception on ***L152*** while running my policies tests
`loc := e.Location.File` anyway to protect it?

here is [link](https://github.com/aquasecurity/trivy-operator/actions/runs/8957717099/job/24601151551?pr=2054) for test failure

Comment (nikpivkin) [reply pkg/iac/rego/load.go:152]: Location can be nil. I'll add a check.

Comment (nikpivkin) [reply pkg/iac/rego/load.go:152]: @chen-keinan Opened the PR with a fix. https://github.com/aquasecurity/trivy/pull/6638

**Pull request merged**